### PR TITLE
Upgrade spl-token to v0.3.8

### PR DESCRIPTION
**Pull request by odcheung**
https://app.asana.com/0/1202413563699470/1204278477629625/f

- upgrade spl-token to v0.3.8
- update JS target to ES2020 to support bigint literals
- Update jest layer for ES upgrade
- Convert all u64 references to BN
- Remove all deprecated `Token` methods, replaced with newer alternatives
- Rework account request util methods 
- (Pending) token-sdk changes

## Breaking Changes
- u64 rename / type change -`Percentage`, `DecimalUtil`, `Ata-utils`
- removed `deriveATA` - use spl-token's `getAssociatedTokenAddressSync` instead
- removed `TokenUtil` functions to use better spl-token alternatives
- Rename `createWSOLAccountInstructions ` to `TokenUtil.createWrappedNativeAccountInstruction`
- EntityParsing now takes in an additional `address: PublicKey` field during parsing
- Updated API for account-request methods `getMultipleAccounts`, `getMultipleAccountsInMap`, `getMultipleParsedAccounts`

## Verification
- Unit-tests on common-sdk
- Updated & ran the unit-tests whirlpools-sdk that heavily relies on the features here

### CODE REVIEW

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 1]
```diff
@@ -28,13 +26,13 @@ describe("ata-util", () => {
...
     // verify address & instruction
-    const notExpected = await tokenGetATA(wallet.publicKey, NATIVE_MINT);
+    const notExpected = getAssociatedTokenAddressSync(wallet.publicKey, NATIVE_MINT);
     const resolved = await resolveOrCreateATA(
       connection,
       wallet.publicKey,
       NATIVE_MINT,
       () => connection.getMinimumBalanceForRentExemption(AccountLayout.span),
-      new u64(LAMPORTS_PER_SOL),
+      BigInt(LAMPORTS_PER_SOL),
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:35]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 2]
```diff
@@ -50,13 +48,13 @@ describe("ata-util", () => {
...
-    const expected = await tokenGetATA(wallet.publicKey, mint.publicKey);
+    const expected = getAssociatedTokenAddressSync(mint, wallet.publicKey);
     const resolved = await resolveOrCreateATA(
       connection,
       wallet.publicKey,
-      mint.publicKey,
+      mint,
       () => connection.getMinimumBalanceForRentExemption(AccountLayout.span),
-      new u64(0),
+      0n,
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:57]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 3]
```diff
@@ -79,17 +77,17 @@ describe("ata-util", () => {
...
 
     // verify address & instruction
     const resolved = await resolveOrCreateATA(
       connection,
       wallet.publicKey,
-      mint.publicKey,
+      mint,
       () => connection.getMinimumBalanceForRentExemption(AccountLayout.span),
-      new u64(0),
+      0n,
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:90]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 4]
```diff
@@ -100,13 +98,13 @@ describe("ata-util", () => {
...
-    const expected = await tokenGetATA(wallet.publicKey, mint.publicKey);
+    const expected = getAssociatedTokenAddressSync(mint, wallet.publicKey);
     const resolved = await resolveOrCreateATA(
       connection,
       wallet.publicKey,
-      mint.publicKey,
+      mint,
       () => connection.getMinimumBalanceForRentExemption(AccountLayout.span),
-      new u64(0),
+      0n,
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:107]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 5]
```diff
@@ -128,13 +126,13 @@ describe("ata-util", () => {
...
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:135]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 6]
```diff
@@ -143,7 +141,7 @@ describe("ata-util", () => {
     expect(resolved.instructions[0].data[0]).toEqual(1); // 1 byte data
 
     // created before execution
-    await mint.createAssociatedTokenAccount(wallet.publicKey);
+    await createAssociatedTokenAccount(connection, wallet.payer, mint, wallet.publicKey)
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:144]: super nit: ; is missing (prettier doesn't fix it ?)

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 7]
```diff
@@ -157,15 +155,13 @@ describe("ata-util", () => {
     const mints = await Promise.all([createNewMint(ctx), createNewMint(ctx), createNewMint(ctx)]);
 
     // create first ATA
-    await mints[0].createAssociatedTokenAccount(wallet.publicKey);
+    await createAssociatedTokenAccount(connection, wallet.payer, mints[0], wallet.publicKey)
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:158]: super nit: ; is missing (prettier doesn't fix it ?)

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 8]
```diff
@@ -180,7 +176,7 @@ describe("ata-util", () => {
     expect(resolved[2].instructions[0].data.length).toEqual(0); // no instruction data
 
     // create second ATA before execution
-    await mints[1].createAssociatedTokenAccount(wallet.publicKey);
+    await createAssociatedTokenAccount(connection, wallet.payer, mints[1], wallet.publicKey)
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:179]: super nit: ; is missing (prettier doesn't fix it ?)

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 9]
```diff
@@ -220,7 +215,7 @@ describe("ata-util", () => {
     expect(resolved[2].instructions[0].data[0]).toEqual(1); // 1 byte data
 
     // create second ATA before execution
-    await mints[1].createAssociatedTokenAccount(wallet.publicKey);
+    await createAssociatedTokenAccount(connection, wallet.payer, mints[1], wallet.publicKey)
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:218]: super nit: ; is missing (prettier doesn't fix it ?)

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 10]
```diff
@@ -157,15 +155,13 @@ describe("ata-util", () => {
...
 
-    const expected = await Promise.all(
-      mints.map((mint) => tokenGetATA(wallet.publicKey, mint.publicKey))
-    );
+    const expected = mints.map((mint) => getAssociatedTokenAddressSync(mint, wallet.publicKey))
     const resolved = await resolveOrCreateATAs(
       connection,
       wallet.publicKey,
-      mints.map((mint) => ({ tokenMint: mint.publicKey, wrappedSolAmountIn: new u64(0) })),
+      mints.map((mint) => ({ tokenMint: mint, wrappedSolAmountIn: 0n })),
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:164]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/ata-util.test.ts [hunk 11]
```diff
@@ -197,15 +193,14 @@ describe("ata-util", () => {
...
+    const expected = mints.map((mint) => getAssociatedTokenAddressSync(mint, wallet.publicKey))
 
-    const expected = await Promise.all(
-      mints.map((mint) => tokenGetATA(wallet.publicKey, mint.publicKey))
-    );
     const resolved = await resolveOrCreateATAs(
       connection,
       wallet.publicKey,
-      mints.map((mint) => ({ tokenMint: mint.publicKey, wrappedSolAmountIn: new u64(0) })),
+      mints.map((mint) => ({ tokenMint: mint, wrappedSolAmountIn: 0n })),
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/ata-util.test.ts:203]: I think resolveOrCreateATA receives lamports in BN. Is BigInt also acceptable ?

File: packages/common-sdk/tests/web3/network/account-requests.test.ts [hunk 12]
```diff
@@ -12,47 +12,53 @@ jest.setTimeout(100 * 1000 /* ms */);
...
 
   it("getParsedAccount, ok", async () => {
     const mint = await createNewMint(ctx);
-    const expected = await mint.getMintInfo();
-    const mintInfo = await getParsedAccount(ctx.connection, mint.publicKey, ParsableMintInfo);
-    expectMintInfoEquals(mintInfo!, expected);
+    const expected = await getMint(ctx.connection, mint);
+
+    const mintInfo = await getParsedAccount(ctx.connection, mint, ParsableMintInfo);
+    expectMintEquals(mintInfo!.data, expected);
```
Comment (yugure-orca) [inline packages/common-sdk/tests/web3/network/account-requests.test.ts:29]: When I run ``yarn run test --verbose``, I found the following error. 🤔
```
tests/web3/network/account-requests.test.ts:29:32 - error TS2339: Property 'data' does not exist on type 'Mint'.

    29     expectMintEquals(mintInfo!.data, expected);
```
There were same errors in this file.

**Review by yugure-orca — APPROVED** [review]
> Map and Record are similar but Map has cleaner syntax and the option to maintain orderliness. This can be very useful for fetcher. User can pass in an array of addresses and we return a map while giving them an option to fetch by index of their original input array.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

Now I understand that there was a clear intent to make the change.
The fact that Map preserves insertion order was a surprise to me, but may be useful once I get used to it.

File: packages/common-sdk/src/web3/network/account-requests.ts [hunk 13]
```diff
@@ -23,36 +24,71 @@ export async function getMultipleParsedAccounts<T>(
...
+    return parser.parse(val[0], val[1]);
+  });
   invariant(results.length === addresses.length, "not enough results fetched");
   return results;
 }
 
-type GetMultipleAccountsInfoResponse = (AccountInfo<Buffer> | null)[];
+// An entry between the key of an address and the account data for that address.
+export type FetchedAccountEntry = [PublicKey, AccountInfo<Buffer> | null];
+export type FetchedAccountMap = Map<string, AccountInfo<Buffer> | null>;
```
[the remark this answers is not in the document]
Comment (odcheung) [reply packages/common-sdk/src/web3/network/account-requests.ts:45]: Discussed offline - we'll be slowly moving towards Map when appropriate. In this scenario (& many AccountFetcher instances), the Map property to preserve orderliness is desired.

**Pull request merged**